tools/clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -3729,6 +3730,13 @@ VisitAbstractConditionalOperator(const AbstractConditionalOperator *E) {
           CGF, E, LHS->getType(), {Cond, LHS, RHS});
     }
   }
+
+  llvm::Value *ResultAlloca = nullptr;
+  if (CGF.getLangOpts().HLSL && CGF.getLangOpts().EnableShortCircuit &&
+      hlsl::IsHLSLMatType(E->getType())) {
+    llvm::Type *MatTy = CGF.ConvertType(E->getType());
+    ResultAlloca = Builder.CreateAlloca(MatTy);
```

Comment:
We previously, perhaps superstitiously, tried to push all the allocas at the top of the code block. Is this going to maintain that? Need it?

elsewhere, we've used `dxilutil::FindAllocaInsertionPoint(F)`, which would also allow this operation to be done below where the alloca is used

Reply:
At the very least, allocas must be inserted in the entry block.  You should use `CGF.CreateTempAlloca` instead, that will place it correctly.

Reply:
Alsok, shouldn't we do this after the `isCheapEnoughToEvaluateUnconditionally` part when we know we'll use it?